src/UI/Keybindings.hs

```diff
@@ -1,35 +1,66 @@
+{-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE OverloadedStrings #-}
 module UI.Keybindings where
 
 import qualified Brick.Main as M
 import qualified Brick.Types as T
 import qualified Brick.Widgets.Edit as E
+import qualified Brick.Widgets.List as L
 import Control.Lens.Getter (view)
+import Control.Lens.At (at)
 import Control.Lens.Setter (set)
+import Control.Lens.Lens (Lens')
+import Control.Monad.IO.Class (liftIO)
 import Data.List (find)
+import Data.Maybe (fromMaybe)
+import Data.Text (Text)
 import Graphics.Vty.Input.Events (Event)
 import Prelude hiding (readFile, unlines)
+import UI.Index.Keybindings
+       (updateStateWithParsedMail, updateReadState)
+import Storage.Notmuch (removeTag)
 import Types
 
 
--- | A generic event handler using Keybindings by default if available
-handleEvent
-    :: [Keybinding]  -- ^ Keybindings to lookup
-    -> (AppState -> Event -> T.EventM Name (T.Next AppState))  -- ^ default handler if no keybinding matches
-    -> AppState
-    -> T.BrickEvent Name e
-    -> T.EventM Name (T.Next AppState)
-handleEvent kbs def s (T.VtyEvent ev) =
-    case lookupKeybinding ev kbs of
-        Just kb -> view kbAction kb s
-        Nothing -> def s ev
-handleEvent _ _ s _ = M.continue s
+-- | try look up a registered keybinding in the configuration and if that fails
+--   redirect the event to a core widget selected for the current application mode
+handleKeyboardEvent :: AppState
+                    -> T.BrickEvent Name e
+                    -> T.EventM Name (T.Next AppState)
+handleKeyboardEvent s (T.VtyEvent ev) =
+    let mode = view asAppMode s
+        kbs = fromMaybe [] $ view (asConfig . confKeybindings . at mode) s
+    in case lookupKeybinding ev kbs of
+           Just kb -> view kbAction kb s
+           Nothing -> fallbackKeybindingHandler mode s ev
+handleKeyboardEvent s _ = M.continue s
+
+
+fallbackKeybindingHandler :: Mode
```

Comment:
So what this simplifies is the overall registration of the key bindings into a central place. What I can not get around is how I should model some kind of fall through mode (like in pattern matching how something matches all). I do this here with this `fallbackKeybindingHandler` but it simply attributes for the fact, that some keys should be allowed to be forwarded to the widget.

Simply put, all keybindings work such that we forward all keys to the widget which has the focus by the mode (currently hard coded), except what is defined in the keybinding.

For example the editor is a prime example: all keys should be forwarded to the editor (since we want to type something), except the ones defined in the key bindings. Is this way perhaps acceptable or do we need something better? At least it seems the most expected when re-configuring the keybindings in the configuration.

Also, looking at #73 I suppose just handling the action by a keybinding is not enough. We need something which can come after the action. Arguably this could be looked at when working on #73 .